### ISSUE TITLE
lost messages

### ISSUE THREAD

**Issue by sjrsrtj** [gkdr/lurch#57]
I set up Lurch with a buddy, everything works just fine. But when my buddy switch to another XMPP-Client without OMEMO support, all messages i sent get lost without further notification. The windowtitle still says (OMEMO) and it lookls like everything works fine. With Conversations my buddy would get this meesage:

>   I sent you an OMEMO encrypted message but your client doesn’t seem to support that. 
  Find more information on https://conversations.im/omemo

It would be great if somehow Lurch could also notify us or our chat partner that somthing went wrong.

Thanks

**Comment by gkdr** [comment]
I switched off sending an additional `<body>` like this because ChatSecure can't handle those (see [here](https://github.com/ChatSecure/ChatSecure-iOS/issues/717)).
Instead, I send an [EME](https://xmpp.org/extensions/xep-0380.html) element, and as far as I know, Conversations switched to that also because the additional `<body>` seems like a dirty hack.

Which client does your friend use? Maybe handling of EME should be requested there.

In any case, the message doesn't get "lost", it just doesn't make sense to send a message both encrypted and in plaintext if one of the devices doesn't support encryption, so it gets sent as encrypted and your friend's client doesn't know what to do with it.
Since there is no mapping of XMPP resource and device ID in OMEMO, the best I could do on the sending side is to issue a warning if the number of resources is not equal to the number of devices on the device list.
I'll try to find out how other clients handle this case.

**Comment by sjrsrtj** [comment]
Ok, so technically messgages didn't get lost. But anyways you could't see them, or a warning, unless you look at your client logs. Which means that they are practically lost.
Same behaviour with Gajim and Pidgin (both with OMEMO deactivated). So should i report it there as well?

**Comment by gkdr** [comment]
I was just trying to say that there is in fact a message to notify clients, it just isn't parsed because the clients don't support that XEP.
If I do it the old way, ChatSecure becomes unusable because it only displays the warning message, but not the actual message.

Yes, this is a feature that should be requested from the developers of a client - I don't think it makes much sense to develop a plugin for this XEP - who would install a plugin just for warning messages?
Alternatively, you can go to the ChatSecure issue I linked and ask for a fix, then I could turn the `<body>` back on.

As I said, I'll try to figure out a way to at least display it to the sender.

**Comment by sjrsrtj** [comment]
> who would install a plugin just for warning messages?

no one would :)

I created a ticket for Pidgin.
[https://developer.pidgin.im/ticket/17208](url)

Thanks for your quick reply

**Comment by GigabyteProductions** [comment]
It appears that backwards compatibility is a _major_ goal of a lot of the XEPs. If a client doesn't implement a XEP properly (which is _not_ the same thing as a client not implementing a XEP at all), it is their responsibility to fix, not everyone else's. Most of the XEPs seem to have specifications that allow unsupporting clients to get something useful (such as [message edits](https://xmpp.org/extensions/xep-0308.html) also sending the new version _as a body tag_ for clients not implementing this XEP).

@gkdr, you even noted in the ChatSecure [issue](https://github.com/ChatSecure/ChatSecure-iOS/issues/717#issuecomment-285907657) that the EME XEP even _suggests_ a kind of body tag for _exactly this issue_. For others. see https://xmpp.org/extensions/xep-0380.html#rules:
> Entities SHOULD include a non-encrypted body as possible, since older clients not supporting this protocol might otherwise ignore messages sent with an unknown encryption, making both the sender frustrated that their message did not get an answer, and the recipient frustrated that they never saw any message.

Sure, a "ugly" body tag may seem weird, but not complying with the specification's recommendation is already proving problematic and breaks (backwards) compatibility worse than compatibility with ChatSecure (clients receiving nothing rather than receiving an indication that something is wrong).

Working around noncompliance is not a good solution.

**Comment by GigabyteProductions** [comment]
Side note: Does anybody know why Conversations can successfully send messages to ChatSecure with OMEMO without dropping the "I sent you an OMEMO encrypted message but your client doesn’t seem to support that. Find more information on https://conversations.im/omemo" ?

**Comment by GigabyteProductions** [comment]
Was also frustrating in testing for https://github.com/gkdr/lurch/issues/63.

**Comment by gkdr** [comment]
You are saying working around noncompliance is not a good solution, but all the other people asking me why they cannot talk to ChatSecure don't.
I think it's a very easy fix and you should bother someone from ChatSecure about it, as they said they are going to implement it anyway (over at the other linked issue). Until then, compatibility is more important.

**Comment by mejo-** [comment]
I agree that messages to OMEMO > Since there is no mapping of XMPP resource and device ID in OMEMO, the best I could do on the sending side is to issue a warning if the number of resources is not equal to the number of devices on the device list.

That would be very helpful. I think that silently dropped messages (and from a UX perspective, they are lost) should be avoided by all means.
